boa/src/builtins/array/mod.rs

```diff
@@ -487,6 +488,45 @@ impl Array {
         Ok(a.into())
     }
 
+    ///'Array.prototype.at(index)'
+    ///
+    /// The at() method takes an integer value and returns the item at that
+    /// index, allowing for positive and negative integers. Negative integers
+    /// count back from the last item in the array.
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///  - [MDN documentation][mdn]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-array.prototype.at
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at
+    pub(crate) fn at(this: &JsValue, args: &[JsValue], context: &mut Context) -> JsResult<JsValue> {
+        //1. let O be ? ToObject(this value)
+        let obj = this.to_object(context)?;
+        //2. let len be ? LengthOfArrayLike(O)
+        let len = obj.length_of_array_like(context)? as i64;
+        //3. let relativeIndex be ? ToIntegerOrInfinity(index)
+        let relative_index = args.get_or_undefined(0).to_integer_or_infinity(context)?;
+        let k = match relative_index {
+            //4. if relativeIndex >= 0, then let k be relativeIndex
+            IntegerOrInfinity::Integer(i) if i >= 0 => i,
+            //5. Else, let k be len + relativeIndex
+            IntegerOrInfinity::Integer(i) => len + i,
```

Comment:
Now that I think about it, you can just use the conditions as match guards for the cases:
```suggestion
            IntegerOrInfinity::Integer(i) if i >= 0 && i < len => i,
            //5. Else, let k be len + relativeIndex
            IntegerOrInfinity::Integer(i) if i.abs() <= len => len + i,
```

That would remove the need for an if to check the validity of k.

Reply:
Whoa...that's so true...

I'll definitely add that. Match guards are so wild sometimes.

I'm going to try to add comments to note exactly the checks in case anybody else needs to read them. Hopefully, they are make sense.

Reply:
Actually I think there is a small potential bug/logic gap in the above match guards.

If the input argument is len, then it would miss the first arm but trigger the second (returning essentially len + len). It looks like obj.get(k) does cover for this gap by returning undefined, but I'm making a small change to the match guard on the second arm to explicitly check that i != len.

Reply:
Ah missed that. You can, then, check for `i < 0 && i.abs() <= len`, which describes better the condition than `i != len && i.abs() <= len`